### ISSUE TITLE
Pier creation problem with `make DEBUG=yes`

### ISSUE THREAD

**Issue by jfranklin9000** [urbit/urbit#844]
Mac OS X El Capitan -- maintenance v0.4.5

````
make DEBUG=yes
bin/urbit -c mycomet
````

results in an endless loop:

````
recover: dig: over
over
/~zod/home/~2017.5.17..03.58.49..c957/arvo/hoon:<[11.103 15].[11.134 57]>
/~zod/home/~2017.5.17..03.58.49..c957/arvo/hoon:<[11.104 15].[11.134 57]>
/~zod/home/~2017.5.17..03.58.49..c957/arvo/hoon:<[11.105 15].[11.134 57]>
/~zod/home/~2017.5.17..03.58.49..c957/arvo/hoon:<[11.107 15].[11.134 57]>
/~zod/home/~2017.5.17..03.58.49..c957/arvo/hoon:<[11.109 15].[11.134 57]>
/~zod/home/~2017.5.17..03.58.49..c957/arvo/hoon:<[11.109 29].[11.109 44]>
/~zod/home/~2017.5.17..03.58.49..c957/arvo/hoon:<[11.174 3].[11.179 13]>
/~zod/home/~2017.5.17..03.58.49..c957/arvo/hoon:<[11.176 3].[11.179 13]>
/~zod/home/~2017.5.17..03.58.49..c957/arvo/hoon:<[11.177 3].[11.179 13]>
/~zod/home/~2017.5.17..03.58.49..c957/arvo/hoon:<[11.178 5].[11.178 47]>
[%swim-call-vane %a %soft ~]
/~zod/home/~2017.5.17..03.58.49..c957/arvo/ames:<[1.567 7].[1.579 17]>
/~zod/home/~2017.5.17..03.58.49..c957/arvo/ames:<[1.576 7].[1.579 17]>
...
...
````

````
^z
````

````
make clean ; make
bin/urbit mycomet
````

finishes the comet creation (almost immediately).

````
make clean ; make
bin/urbit -c mycomet2
````

also works.

This is repeatable for me with my current system.  I haven't tried it on linux.

I'll try to ID the offending file.

**Comment by juped** [comment]
Tried it on Debian, didn't happen there.

**Comment by jfranklin9000** [comment]
The problem file is `noun/nock.c` if it's *not* compiled with optimization.

If you compile and link everything with `-O0` it will `dig: over` on ship creation.

If you then recompile `noun/nock.c` with `-O1` (still link with `-O0`) it will complete ship creation.

````
cc -v
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
````

**Comment by cgyarvin** [comment]
Wow!  That's really interesting.  I thought everything nock.c does was perfectly TCOed just by the rules of C.  Maybe it's just a deep call stack and poorly handled spilling...

**Comment by jfranklin9000** [comment]
If I `#define NO_OVERFLOW` in `noun/manage.c` I get a SIGSEGV with a huge backtrace:

````
Thread 1 received signal SIGSEGV, Segmentation fault.
0x0000000100012b79 in _cj_fine (cup=3227634141, mop=3228065785, cor=3221868594) at noun/jets.c:542
542	      if ( c3n == u3r_sing(u3t(par), u3t(cor)) ) {
(gdb) bt 20
#0  0x0000000100012b79 in _cj_fine (cup=3227634141, mop=3228065785, cor=3221868594) at noun/jets.c:542
#1  0x0000000100013393 in _cj_fine (cup=3227634141, mop=3228065785, cor=3221868594) at noun/jets.c:569
#2  0x0000000100013393 in _cj_fine (cup=3230744982, mop=3231798876, cor=3221928987) at noun/jets.c:569
#3  0x0000000100013393 in _cj_fine (cup=3223199924, mop=3230924052, cor=3225530235) at noun/jets.c:569
#4  0x0000000100011f18 in u3j_kick (cor=3758089166, axe=2) at noun/jets.c:621
#5  0x000000010001bca2 in _n_nock_on (bus=3758089214, fol=3221844816) at noun/nock.c:339
#6  0x000000010001b69e in _n_nock_on (bus=3758089046, fol=3224257899) at noun/nock.c:204
#7  0x000000010001b6c1 in _n_nock_on (bus=3758089046, fol=3224257844) at noun/nock.c:205
#8  0x000000010001bc29 in _n_nock_on (bus=3758089046, fol=3223459429) at noun/nock.c:335
#9  0x000000010001b69e in _n_nock_on (bus=3758089088, fol=3223794112) at noun/nock.c:204
#10 0x000000010001b6c1 in _n_nock_on (bus=3758088992, fol=3225851572) at noun/nock.c:205
#11 0x000000010001b69e in _n_nock_on (bus=3758088992, fol=3225256019) at noun/nock.c:204
#12 0x000000010001b6c1 in _n_nock_on (bus=3758088992, fol=3225851560) at noun/nock.c:205
#13 0x000000010001bc29 in _n_nock_on (bus=3758088992, fol=3225654798) at noun/nock.c:335
#14 0x000000010001b6c1 in _n_nock_on (bus=3758089058, fol=3225863844) at noun/nock.c:205
#15 0x000000010001b6c1 in _n_nock_on (bus=3758089058, fol=3224965441) at noun/nock.c:205
#16 0x000000010001b6c1 in _n_nock_on (bus=3758089058, fol=3225863820) at noun/nock.c:205
#17 0x000000010001b69e in _n_nock_on (bus=3758089058, fol=3225845817) at noun/nock.c:204
#18 0x000000010001b6c1 in _n_nock_on (bus=3758089058, fol=3225863808) at noun/nock.c:205
#19 0x000000010001b6c1 in _n_nock_on (bus=3758089058, fol=3225863796) at noun/nock.c:205
(gdb) bt -50
#144575 0x000000010001b69e in _n_nock_on (bus=3758095079, fol=3223946472) at noun/nock.c:204
#144576 0x000000010001b6c1 in _n_nock_on (bus=3758095079, fol=3224710332) at noun/nock.c:205
#144577 0x000000010001c7dd in _n_hint (zep=1851876717, hod=3758095434, bus=3758095482, nex=3224717457) at noun/nock.c:70
#144578 0x000000010001bfb3 in _n_nock_on (bus=3758095482, fol=3223139573) at noun/nock.c:389
#144579 0x000000010001bbbe in _n_nock_on (bus=3758095482, fol=3224529811) at noun/nock.c:318
#144580 0x000000010001bbbe in _n_nock_on (bus=3758095590, fol=3224574765) at noun/nock.c:318
#144581 0x000000010001bbbe in _n_nock_on (bus=3758095584, fol=3224019981) at noun/nock.c:318
#144582 0x000000010001bbbe in _n_nock_on (bus=3758095344, fol=3224019981) at noun/nock.c:318
#144583 0x000000010001bbbe in _n_nock_on (bus=3758095596, fol=3224019981) at noun/nock.c:318
#144584 0x000000010001bbbe in _n_nock_on (bus=3758095608, fol=3224660878) at noun/nock.c:318
#144585 0x000000010001c7dd in _n_hint (zep=1953460339, hod=3224745178, bus=3758095746, nex=3224746930) at noun/nock.c:70
#144586 0x000000010001bfb3 in _n_nock_on (bus=3758095746, fol=3224483641) at noun/nock.c:389
#144587 0x000000010001bbbe in _n_nock_on (bus=3758095746, fol=3224050207) at noun/nock.c:318
#144588 0x000000010001c7dd in _n_hint (zep=1953460339, hod=3223574423, bus=3758095746, nex=3223457119) at noun/nock.c:70
#144589 0x000000010001bfb3 in _n_nock_on (bus=3758095746, fol=3223424328) at noun/nock.c:389
#144590 0x000000010001c7dd in _n_hint (zep=1953460339, hod=3224744834, bus=3758095746, nex=3223424328) at noun/nock.c:70
#144591 0x000000010001bfb3 in _n_nock_on (bus=3758095746, fol=3224505533) at noun/nock.c:389
#144592 0x000000010001c7dd in _n_hint (zep=1953460339, hod=3224743092, bus=3758095788, nex=3224636557) at noun/nock.c:70
#144593 0x000000010001bfb3 in _n_nock_on (bus=3758095788, fol=3224535786) at noun/nock.c:389
#144594 0x000000010001c7dd in _n_hint (zep=1953460339, hod=3224875917, bus=3758095884, nex=3223135092) at noun/nock.c:70
#144595 0x000000010001bfb3 in _n_nock_on (bus=3758095884, fol=3224875911) at noun/nock.c:389
#144596 0x000000010001bbbe in _n_nock_on (bus=3758095884, fol=3224876379) at noun/nock.c:318
#144597 0x000000010001c7dd in _n_hint (zep=1953460339, hod=3224152662, bus=3758095884, nex=3224875769) at noun/nock.c:70
#144598 0x000000010001bfb3 in _n_nock_on (bus=3758095884, fol=3224875684) at noun/nock.c:389
#144599 0x000000010001c7dd in _n_hint (zep=1953460339, hod=3224190040, bus=3758095884, nex=3224873892) at noun/nock.c:70
#144600 0x000000010001bfb3 in _n_nock_on (bus=3758095884, fol=3224190068) at noun/nock.c:389
#144601 0x000000010001c7dd in _n_hint (zep=1953460339, hod=3224872074, bus=3758095884, nex=3223208066) at noun/nock.c:70
#144602 0x000000010001bfb3 in _n_nock_on (bus=3758095884, fol=3223277587) at noun/nock.c:389
#144603 0x000000010001c7dd in _n_hint (zep=1953460339, hod=3224733849, bus=3758095866, nex=3223701229) at noun/nock.c:70
#144604 0x000000010001bfb3 in _n_nock_on (bus=3758095866, fol=3224869588) at noun/nock.c:389
#144605 0x000000010001c7dd in _n_hint (zep=1953460339, hod=3224869447, bus=3758095866, nex=3224869588) at noun/nock.c:70
#144606 0x000000010001bfb3 in _n_nock_on (bus=3758095866, fol=3224826715) at noun/nock.c:389
#144607 0x000000010001b505 in u3n_nock_on (bus=3758095866, fol=3758095860) at noun/nock.c:436
#144608 0x0000000100013629 in u3j_kink (cor=3758095866, axe=2) at noun/jets.c:676
#144609 0x000000010001c448 in u3n_kick_on (gat=3758095866) at noun/nock.c:447
#144610 0x000000010001c5dd in u3n_slam_on (gat=3758095872, sam=3758095884) at noun/nock.c:468
#144611 0x00000001000256b6 in _cv_nock_poke (ovo=3230568023) at noun/vortex.c:226
#144612 0x0000000100025533 in u3v_poke (ovo=3230568023) at noun/vortex.c:397
#144613 0x00000001000185df in u3m_soft_top (sec_w=600, pad_w=1048576, fun_f=0x100025520 <u3v_poke>, arg=3230568023) at noun/manage.c:901
#144614 0x00000001000195ab in u3m_soft (sec_w=600, fun_f=0x100025520 <u3v_poke>, arg=3230568023) at noun/manage.c:1156
#144615 0x0000000100069fa9 in _raft_punk (ovo=3230568023) at vere/raft.c:1574
#144616 0x00000001000692b5 in u3_raft_work () at vere/raft.c:1991
#144617 0x00000001000675b3 in u3_lo_shut (inn=0 '\000') at vere/loop.c:449
#144618 0x000000010005ea15 in _ames_recv_cb (wax_u=0x100123268 <u3_Host+776>, nrd_i=726,
    buf_u=0x7fff5fbf7810, adr_u=0x7fff5fbf7850, flg_i=0) at vere/ames.c:304
#144619 0x0000000100092004 in uv__udp_recvmsg (handle=<optimized out>) at src/unix/udp.c:201
#144620 uv__udp_io (loop=<optimized out>, w=<optimized out>, revents=<optimized out>) at src/unix/udp.c:139
#144621 0x0000000100094aff in uv__io_poll (loop=<optimized out>, timeout=<optimized out>) at src/unix/kqueue.c:247
#144622 0x0000000100085a48 in uv_run (loop=<optimized out>, mode=UV_RUN_DEFAULT) at src/unix/core.c:341
#144623 0x0000000100067815 in u3_lo_loop () at vere/loop.c:660
#144624 0x000000010005c1c9 in main (argc=2, argv=0x7fff5fbffc38) at vere/main.c:572
````

(This is not so clear using `Sigstk[]` (`#undef NO_OVERFLOW`).)

[https://clang.llvm.org/docs/UsersManual.html#options-to-emit-optimization-reports](https://clang.llvm.org/docs/UsersManual.html#options-to-emit-optimization-reports)

````
cc -Rpass=.* (...) -O0 (...) -o noun/nock.o noun/nock.c
````
*no output*

````
cc -Rpass=.* (...) -O1 (...) -o noun/nock.o noun/nock.c

noun/nock.c:166:13: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      ite = u3nc(c3__leaf, u3_nul);
            ^
noun/nock.c:170:12: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
    return u3nc(ite, _n_mush_in(u3t(val)));
           ^
noun/nock.c:170:12: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:168:13: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      ite = u3nc(c3__leaf, u3qe_trip(h_val));
            ^
noun/nock.c:168:13: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:181:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  pro = u3nt(c3__rose,
        ^
noun/nock.c:181:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:181:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:181:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:181:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:184:3: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  u3z(val);
  ^
noun/nock.c:142:7: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3z(hod);
      ^
noun/nock.c:144:23: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3_noun pro = _n_nock_on(bus, nex);
                      ^
noun/nock.c:117:7: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3z(hod);
      ^
noun/nock.c:122:23: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3_noun pro = u3z_find_2(c3__nock, bus, nex);
                      ^
noun/nock.c:125:11: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
          u3z(bus); u3z(nex);
          ^
noun/nock.c:125:21: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
          u3z(bus); u3z(nex);
                    ^
noun/nock.c:128:26: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        pro = _n_nock_on(u3k(bus), u3k(nex));
                         ^
noun/nock.c:128:36: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        pro = _n_nock_on(u3k(bus), u3k(nex));
                                   ^
noun/nock.c:128:15: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        pro = _n_nock_on(u3k(bus), u3k(nex));
              ^
noun/nock.c:131:11: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
          u3z_save_2(c3__nock, bus, nex, pro);
          ^
noun/nock.c:134:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3z(bus); u3z(nex);
        ^
noun/nock.c:134:19: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3z(bus); u3z(nex);
                  ^
noun/nock.c:107:21: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3_noun pro = _n_nock_on(bus, nex);
                    ^
noun/nock.c:109:7: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3t_off(noc_o);
      ^
noun/nock.c:110:21: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3j_mine(hod, u3k(pro));
                    ^
noun/nock.c:110:7: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3j_mine(hod, u3k(pro));
      ^
noun/nock.c:111:7: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3t_on(noc_o);
      ^
noun/nock.c:97:21: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3_noun pro = _n_nock_on(bus, nex);
                    ^
noun/nock.c:99:19: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      if ( c3y == u3r_sing(pro, hod) ) {
                  ^
noun/nock.c:102:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3z(hod); return pro;
        ^
noun/nock.c:100:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3z(pro); return hod;
        ^
noun/nock.c:93:14: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      return _n_nock_on(bus, nex);
             ^
noun/nock.c:89:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3t_off(noc_o);
        ^
noun/nock.c:90:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3t_slog(hod);
        ^
noun/nock.c:91:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3t_on(noc_o);
        ^
noun/nock.c:82:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3z(hod);
        ^
noun/nock.c:84:14: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      return _n_nock_on(bus, nex);
             ^
noun/nock.c:78:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3t_off(noc_o);
        ^
noun/nock.c:79:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3t_heck(hod);
        ^
noun/nock.c:80:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
        u3t_on(noc_o);
        ^
noun/nock.c:47:21: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3_noun tac = u3nc(zep, hod);
                    ^
noun/nock.c:50:7: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3t_push(tac);
      ^
noun/nock.c:70:13: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      pro = _n_nock_on(bus, nex);
            ^
noun/nock.c:71:7: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3t_drop();
      ^
noun/nock.c:37:7: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3a_lose(zep);
      ^
noun/nock.c:38:7: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      u3a_lose(hod);
      ^
noun/nock.c:40:14: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
      return _n_nock_on(bus, nex);
             ^
noun/nock.c:435:3: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  u3t_on(noc_o);
  ^
noun/nock.c:436:9: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  pro = _n_nock_on(bus, fol);
        ^
noun/nock.c:437:3: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  u3t_off(noc_o);
  ^
noun/nock.c:447:10: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  return u3j_kink(gat, 2);
         ^
noun/nock.c:457:17: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  u3_noun cor = u3nc(u3k(u3h(gat)), u3nc(sam, u3k(u3t(u3t(gat)))));
                ^
noun/nock.c:457:17: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:457:17: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:457:17: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:457:17: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:457:17: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:467:3: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  u3z(gat);
  ^
noun/nock.c:468:10: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  return u3n_kick_on(cor);
         ^
noun/nock.c:476:10: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  return u3m_soft_run(gul, u3n_nock_on, bus, fol);
         ^
noun/nock.c:484:10: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  return u3m_soft_run(gul, u3n_slam_on, gat, sam);
         ^
noun/nock.c:492:17: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  u3_noun gul = u3nt(u3nt(1, 0, 0), 0, 0);  //  |=(a/{* *} ~)
                ^
noun/nock.c:492:17: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
noun/nock.c:494:10: remark: marked this call a tail call candidate [-Rpass=tailcallelim]
  return u3n_nock_et(gul, bus, fol);
         ^
````

It seems the compiler is doing tail call elimination, saving the stack.

**Comment by jfranklin9000** [comment]
> I think so as it seems the only optimization applicable is tail call elimination:

I'm not sure about this anymore.

I built a version with `-O1` but turned off tail call optimization by adding
`__attribute__((disable_tail_calls))` to all the functions in `nock.c`.
I was expecting it to fail but it didn't, so there could be another
optimization happening that isn't reported by `-Rpass=.*`.

**Comment by jfranklin9000** [comment]
TL;DR

This issue can be mitigated by setting `CFLAGS=-O1 -g` for the `DEBUG=yes` case in `Makefile`.

----

I added this stack depth measuring code (but only measure in `_n_nock_on()`)

````
#### vere/main.c

char *main_stack_address = (char *) 0xffffffffffffffff;
char *nock_stack_address = (char *) 0xffffffffffffffff;
long max_stack_depth = 0;

c3_i
main(c3_i   argc,
     c3_c** argv)
{
  char stack_element = '\0';
  main_stack_address = &stack_element;
..
..
````

````
#### noun/nock.c

static u3_noun
_n_nock_on(u3_noun bus, u3_noun fol)
{
  extern char *main_stack_address;
  extern char *nock_stack_address;
  extern long max_stack_depth;
  char stack_element = '\0';
  char *stack_address = &stack_element;
  if (stack_address < nock_stack_address) {
    nock_stack_address = stack_address;
    max_stack_depth = main_stack_address - nock_stack_address;
  }
..
..
````

Tools

````
cc -v
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
````

````
gcc-7 -v
Using built-in specs.
COLLECT_GCC=gcc-7
COLLECT_LTO_WRAPPER=/usr/local/Cellar/gcc/7.1.0/libexec/gcc/x86_64-apple-darwin15.6.0/7.1.0/lto-wrapper
Target: x86_64-apple-darwin15.6.0
Configured with: ../configure --build=x86_64-apple-darwin15.6.0 --prefix=/usr/local/Cellar/gcc/7.1.0 --libdir=/usr/local/Cellar/gcc/7.1.0/lib/gcc/7 --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-7 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-system-zlib --enable-checking=release --with-pkgversion='Homebrew GCC 7.1.0' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --disable-nls --with-native-system-header-dir=/usr/include --with-sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk
Thread model: posix
gcc version 7.1.0 (Homebrew GCC 7.1.0) 
````

Stack limit

````
ulimit -Hs
65532
````
*64 Mb - 4 Kb*

Compile everything with `clang -O3`

````
make distclean ; make
````

Edit `CFLAGS` in `Makefile` to select different optimization levels for `DEBUG=yes` case.

Use this when selecting `gcc`

````
CC=gcc-7
CXX=g++-7
CXXFLAGS=$(CFLAGS)
CLD=g++-7 $(CFLAGS) -L/usr/local/lib $(OPTLOCALLFLAGS) $(OPENSSLLFLAGS)
````

Build urbit (x6)

````
make clean ; make DEBUG=yes
````

Results

| compiler | `CFLAGS` | `max_stack_depth` | result         |
|:-------- |:-------- | -----------------:|:-------------- |
| `clang`  | `-O0 -g` |      `67.102.948` | stack overflow |
| `clang`  | `-O1 -g` |      `17.285.555` | comet creation |
| `clang`  | `-O3 -g` |      `17.279.771` | comet creation |
| `gcc`    | `-O0 -g` |      `39.460.736` | comet creation |
| `gcc`    | `-O1 -g` |      `14.816.243` | comet creation |
| `gcc`    | `-O3 -g` |      `19.743.615` | comet creation |

Estimate `max_stack_depth` for `clang -O0 -g` by measuring
`max_stack_depth` for `-O0` and `-O3` when booting an
existing comet and assume the ratio is constant

`-O0` = `103.684`

`-O3` = `22.155`

Estimated `max_stack_depth` for `clang -O0 -g`
`= 17279771 * 103684 / 22155 = 77 Mb`

**Comment by Fang-** [comment]
Can still reproduce, now gives a slightly different stack trace (which probably doesn't matter).

```
recover: dig: over
over
[%swim-call-vane %a %soft ~]
/~zod/home/git-e0418a3162/sys/zuse:<[2.672 9].[2.682 52]>
/~zod/home/git-e0418a3162/sys/zuse:<[2.673 9].[2.682 52]>
/~zod/home/git-e0418a3162/sys/zuse:<[2.674 9].[2.682 52]>
/~zod/home/git-e0418a3162/sys/zuse:<[2.675 9].[2.682 52]>
/~zod/home/git-e0418a3162/sys/zuse:<[2.676 9].[2.682 52]>
/~zod/home/git-e0418a3162/sys/zuse:<[2.677 9].[2.682 52]>
/~zod/home/git-e0418a3162/sys/zuse:<[2.678 9].[2.682 52]>
/~zod/home/git-e0418a3162/sys/zuse:<[2.678 13].[2.678 51]>
/~zod/home/git-e0418a3162/sys/zuse:<[2.678 17].[2.678 51]>
```

(This was also observed with debug off, see #936.)

It looks like we maybe already have (the components of) a fix for this here, per @jfranklin9000's comment above?

**Comment by joemfb** [comment]
@Fang-'s error report is the oft-mentioned (and now resolved) former lack of TCO in AES CTR mode.

The original issue (the inability to boot a ship without an optimized binary) remains, but the mitigation is slightly different, as the build system has been replaced.

To create a meson build configuration with debug symbols, the following command should be used:

```
 meson . ./build --buildtype=debugoptimized
```

The binary built from that configuration will be able to boot ships.

**Comment by joemfb** [comment]
With the addition of the bytecode interpreter (which doesn't use the C stack), we can know boot ships without any build optimization. Confirmed with `meson . ./build --buildtype=debug`
